src/main/java/mods/eln/sixnode/EmergencyLamp.kt

```diff
@@ -0,0 +1,320 @@
+            add(Utils.plotVolt("Nominal voltage:", cable.electricalNominalVoltage))
+            add(Utils.plotEnergy("Battery capacity:", batteryCapacity))
+        }
+    }
+}
+
+class EmergencyLampElement(sixNode: SixNode, side: Direction, descriptor: SixNodeDescriptor)
+    : SixNodeElement(sixNode, side, descriptor) {
+
+    enum class Event(val value: Byte) {
+        TOGGLE_POWERED_BY_CABLE(1),
+        SET_CHANNEL(2)
+    }
+
+    val desc = descriptor as EmergencyLampDescriptor
+    val load = NbtElectricalLoad("load")
+    val chargingResistor = ResistorSwitch("chargingResistor", load, null)
+    var on = false
+        set(value) {
+            if (field != value) needPublish()
+            field = value
+            sixNode.lightValue = if (value) desc.lightLevel else 0
+        }
+    var charge = desc.batteryCapacity / 2
+    var poweredByCable = false
+        set(value) {
+            val changed = field != value
+            field = value
+            if (changed) {
+                reconnect()
+                needPublish()
+                if (value) isConnectedToLampSupply = false
+            }
+        }
+
+    var channel by published("Default channel")
+    var isConnectedToLampSupply by published(false)
+
+    val process = IProcess { deltaT ->
+        if (!poweredByCable) {
+            var closestPowerSupply: LampSupplyElement.PowerSupplyChannelHandle? = null
+            var closestDistance = 10000f
+
+            LampSupplyElement.channelMap[channel]?.forEach {
+                val distance = it.element.sixNode.coordonate.trueDistanceTo(sixNode.coordonate).toFloat()
+                if (distance < closestDistance && distance <= it.element.range) {
+                    closestDistance = distance
+                    closestPowerSupply = it
+                }
+            }
+
+            if (closestPowerSupply != null && closestPowerSupply!!.element.getChannelState(closestPowerSupply!!.id)) {
+                isConnectedToLampSupply = true
+                closestPowerSupply!!.element.addToRp(chargingResistor.r)
+                load.state = closestPowerSupply!!.element.powerLoad.state
+            } else {
+                isConnectedToLampSupply = false
+                load.state = 0.0
+            }
+        }
+
+        if (chargingResistor.u > 0.5 * desc.cable.electricalNominalVoltage) {
+            on = false
+            if (charge < desc.batteryCapacity) {
+                chargingResistor.state = true
+                charge = Math.min(charge + chargingResistor.p * deltaT, desc.batteryCapacity)
+            } else {
+                chargingResistor .state = false
+            }
+        } else {
+            chargingResistor.state = false
+            if (charge > 0) {
+                on = true
+                charge = Math.max(charge - desc.consumption * deltaT, 0.0)
+            } else {
+                on = false
+            }
+        }
+    }
+
+    override fun initialize() {
```

Comment:
Honestly, this could all be put in init{}. Doesn't make a difference, I suppose.